### PR TITLE
View Caching

### LINKED ISSUE #1889: View Cache Job tracker

### LINKED ISSUE #1890: Update view handling per caching

### PR DESCRIPTION
**Pull request by drostron**
Closes #1889 
Closes #1890

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -323,6 +323,14 @@ For example, given the above MongoDB mount, an additional view could be defined
 
 A view can be mounted at any file path. If a view's path is nested inside the path of a database mount, it will appear alongside the other files in the database. A view will "shadow" any actual file that would otherwise be mapped to the same path. Any attempt to write data to a view will result in an error.
 
+#### Caching
+
+View mounts can optionally be cached. When cached a view is refreshed periodically in the background with respect to its associated `max-age`.
+
+A cached view is created by adding the `Cache-Control: max-age=<seconds>`  header to a `/mount/fs/` request.
+
+Cached views are retrieved via `/data/fs` and include a `Expires: <http-date>` header informed by `max-age`. A resonablly provisioned server will keep cached views fresh. If it does fall behind it will serve the stale cache with a `110` HTTP status. The `Expires` header is absent from non-cached views.
```
Comment (alissapajer) [inline README.md]: s/resonablly/reasonably/

**Comment by drostron** [comment]
@jedesah Per conversation I'm going to switch `110` to it's proper home as a warning header. Thanks for the observation 😄.

**Comment by drostron** [comment]
@jedesah Ready

File: interface/src/main/scala/quasar/main/Caching.scala [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+  type Eff[A] = (Task :\: ConnectionIO :/: CoreEff)#M[A]
+
+  val QT = QueryFile.Transforms[Free[Eff, ?]]
+
+  final case class ViewCacheRefreshCtx(start: Task[Unit], shutdown: Task[Unit])
+
+  def viewCacheRefreshCtx(eval: Free[Eff, ?] ~> Task): Task[ViewCacheRefreshCtx] = Task.delay {
+    val executor: ScheduledExecutorService =
+      Executors.newScheduledThreadPool(
+        scala.math.max(8, java.lang.Runtime.getRuntime.availableProcessors * 2),
```
Comment (jedesah) [inline interface/src/main/scala/quasar/main/Caching.scala]: What's the rationale for capping at 8?

File: interface/src/main/scala/quasar/main/Caching.scala [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    ViewCacheRefreshCtx(refresh, Task.delay(executor.shutdown))
+  }
+
+  def refresh(
+    f: Free[Eff, ?] ~> Task,
+    assigneeId: String
+  ): Process[Task, Unit] = {
+    val taskToProcess = λ[Task ~> Process[Task, ?]](Process.await(_)(Process.emit))
+
+    taskToProcess(
```
Comment (jedesah) [inline interface/src/main/scala/quasar/main/Caching.scala]: I think `Process.eval` would be more standard here and would avoid the definition of `taskToProcess`

File: interface/src/main/scala/quasar/main/Quasar.scala [hunk 4]
```diff
@@ -85,11 +86,6 @@ object Quasar {
   val absorbTask: QErrsTCnxIO ~> QErrsCnxIO =
     Inject[ConnectionIO, QErrsCnxIO].compose(taskToConnectionIO) :+: reflNT[QErrsCnxIO]
 
-  /** Initialize the Quasar FileSystem assuming all defaults
-    * The metastore can be changed but it will not be persisted to the config file.
-    */
-  val init: MainTask[Quasar] = initFromMetaConfig(None, _ => ().point[MainTask])
-
```
Comment (jedesah) [inline interface/src/main/scala/quasar/main/Quasar.scala]: 😢

File: interface/src/main/scala/quasar/main/Quasar.scala [hunk 5]
```diff
@@ -129,23 +125,39 @@ object Quasar {
...
         (injectFT[Task, QErrs_Task] compose MetaStoreLocation.impl.default(metaRef, persist)) :+:
          injectFT[Task, QErrs_Task]                                                           :+:
         (injectFT[Task, QErrs_Task] compose connectionIOToTask)                               :+:
          injectFT[QErrs, QErrs_Task]
 
-      Quasar(
-        foldMapNT(g) compose foldMapNT(f) compose runCore,
-        (mntdRef.read >>= closeAllFsMounts _) *> metaRef.read.flatMap(_.trans.shutdown))
-    }
+      val h: CoreEff ~> QErrs_TaskM =  foldMapNT(g) compose foldMapNT(f) compose runCore
```
Comment (jedesah) [inline interface/src/main/scala/quasar/main/Quasar.scala]: I can't say that I find the definition of `h` terribly useful but perhaps okay with it if it were to have a more descriptive name

File: interface/src/main/scala/quasar/main/Quasar.scala [hunk 6]
```diff
@@ -129,23 +125,39 @@ object Quasar {
...
+
+      val mainTaskToTask =
+        λ[MainTask ~> Task](_.foldM(e => Task.fail(new RuntimeException(e)), Task.delay(_)))
+
+      val cacheCtx =
+        Caching.viewCacheRefreshCtx(foldMapNT(
+          reflNT[Task]       :+:
+          connectionIOToTask :+:
+          (h andThen QErrs_Task.toMainTask andThen mainTaskToTask)))
```
Comment (jedesah) [inline interface/src/main/scala/quasar/main/Quasar.scala]: Ah, I see it's used here. Still, I think a better name would be nice

File: interface/src/test/scala/quasar/main/CachingSpec.scala [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+final class CachingSpec extends quasar.Qspec with H2MetaStoreFixture {
+  import Caching.Eff
+
+  val schema: quasar.db.Schema[Int] = quasar.metastore.Schema.schema
+
+  type MountingFileSystem[A] = Coproduct[Mounting, FileSystem, A]
+
+  val mount = λ[Mounting ~> Task](_ => Task.fail(new RuntimeException("unimplemented")))
+
+  val transact: ConnectionIO ~> Task = λ[ConnectionIO ~> Task](_.transact(transactor))
```
Comment (jedesah) [inline interface/src/test/scala/quasar/main/CachingSpec.scala]: More easily defined as `transactor.trans`

File: web/src/test/scala/quasar/api/services/DataServiceSpec.scala [hunk 8]
```diff
@@ -320,6 +367,74 @@ class DataServiceSpec extends quasar.Qspec with FileSystemFixture with Http4s {
           }
         }
       }
+
+      "respond with view cache data" >> {
+        "fresh" >> prop {
```
Comment (jedesah) [inline web/src/test/scala/quasar/api/services/DataServiceSpec.scala]: Just as a general comment on this that requires no action. I've been moving away from tests that are too "generic" as then you start running into the problem where the test is just as complicated to understand as the implementation and there is a higher likelihood that both the tests and implementation are not doing what a human would expect to see in manual testing. Obviously, I think property based testing is tremendously useful, but for a test like this I think a few more hardcoded literals might improve the degree of confidence the reader of the test has in the implementation being correct which I think is the ultimate goal.

To summarize, I think that for relatively simple properties, property testing totally increases the degree of confidence in the implementation but if the property is too complex/generic it actually does the opposite.

File: web/src/main/scala/quasar/api/services/mount.scala [hunk 9]
```diff
@@ -123,8 +144,36 @@ object mount {
                  (msg, _) => ApiError.fromMsg_(
                    BadRequest, msg).left))
       exists <- M.lookupType(path).run.isDefined.liftM[ApiErrT]
-      mnt    =  if (replaceIfExists && exists) M.replace(path, bConf)
-                else M.mount(path, bConf)
-      _      <- mnt.liftM[ApiErrT]
+      cc     =  req.headers.get(`Cache-Control`)
```
Comment (jedesah) [inline web/src/main/scala/quasar/api/services/mount.scala]: This stuff should go inside of the `(refineType(path).toOption ⊛ MountConfig.viewConfig.getOption(bConf) ⊛ maxAge ⊛ rAfter)` check. No point computing it if the above isn't satisfied and also I think it would make the code easier to follow despite requiring a nested for comprehension which is a little awkward I will admit

File: web/src/main/scala/quasar/api/services/data.scala [hunk 10]
```diff
@@ -71,16 +83,24 @@ object data {
                     requiredHeader(Destination, req) map (_.value))
         dst    <- EitherT.fromDisjunction[M.FreeS](parseDestination(dstStr))
         scn    <- EitherT.fromDisjunction[M.FreeS](moveScenario(path, dst))
-        _      <- M.move(scn, MoveSemantics.FailIfExists)
```
Comment (jedesah) [inline web/src/main/scala/quasar/api/services/data.scala]: I think this logic should be in the view interpreter

File: web/src/main/scala/quasar/api/services/data.scala [hunk 11]
```diff
@@ -71,16 +83,24 @@ object data {
...
+        _      <- EitherT(vcacheGet(path).fold(sf =>
+                    refineType(scn.dst)
+                      .leftMap(p => PathError.invalidPath(p, "view mount destination must be a file").toApiError)
+                      .traverse(d => VCache.Ops[S].move(sf, d)),
+                    M.move(scn, MoveSemantics.FailIfExists).leftMap(_.toApiError).run).join)
       } yield Created).run)
 
     case DELETE -> AsPath(path) =>
-      respond(M.delete(path).run)
+      respond(
```
Comment (jedesah) [inline web/src/main/scala/quasar/api/services/data.scala]: Same here. (view interpreter)

File: web/src/main/scala/quasar/api/services/mount.scala [hunk 12]
```diff
@@ -77,11 +85,17 @@ object mount {
         })
 
       case DELETE -> AsPath(p) =>
-        respond(M.unmount(p).as(s"deleted ${printPath(p)}"))
+        val deleteViewIfExists: OptionT[Free[S, ?], Unit] =
```
Comment (jedesah) [inline web/src/main/scala/quasar/api/services/mount.scala]: This also would be better if it was implemented in the `unmount` method as opposed to here in order to make it more easily accessible from the Quasar programmatic API without involving a web server

File: web/src/main/scala/quasar/api/package.scala [hunk 13]
```diff
@@ -70,6 +70,8 @@ package object api {
           _.right[Response].point[Task])))
     }
 
+  // https://tools.ietf.org/html/rfc7234#section-4.2.4
```
[the remark this answers is not in the document]
Comment (drostron) [reply web/src/main/scala/quasar/api/package.scala]: Agreed — updated in a subsequent commit per discussion.

File: core/src/main/scala/quasar/fs/cache/ViewCache.scala [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+  object Status {
+    final case object Pending    extends Status
+    final case object Successful extends Status
+    final case object Failed     extends Status
+
+    implicit val equal: Equal[Status] = Equal.equalRef
+  }
+
+  // Hard coded to 80% of maxAge for now
+  def expireAt(ts: Instant, maxAge: Duration): Task[Instant] =
```
[the remark this answers is not in the document]
Comment (drostron) [reply core/src/main/scala/quasar/fs/cache/ViewCache.scala]: `plus` can except

File: core/src/main/scala/quasar/metastore/MetaStoreAccess.scala [hunk 15]
```diff
@@ -48,6 +48,19 @@ trait MetaStoreAccess {
...
     runOneRowUpdateOpt(Queries.deleteMount(path)) toRight NotFound
 
+  //--- View Cache ---
+  val staleCachedViews: ConnectionIO[List[PathedViewCache]] = Queries.staleCachedViews.list
+
+  def lookupViewCache(path: AFile): ConnectionIO[Option[ViewCache]] =
+    Queries.lookupViewCache(path).option ∘ (_ ∘ (_.vc))
+
+  def updateInsertViewCache(path: AFile, viewCache: ViewCache): ConnectionIO[Unit] =
+    runOneRowUpdateOpt(Queries.updateViewCache(path, viewCache)).isEmpty >>= (_.whenM(
```
[the remark this answers is not in the document]
Comment (drostron) [reply core/src/main/scala/quasar/metastore/MetaStoreAccess.scala]: I was originally utilizing PostgreSQL's `INSERT ... ON CONFLICT UPDATE`, but unfortunately H2 doesn't have an equivalent.

[the remark this answers is not in the document]
Comment (drostron) [reply core/src/main/scala/quasar/metastore/MetaStoreAccess.scala]: Ohh, I see, H2 has `ON DUPLICATE KEY UPDATE ...` if running in MySQL mode. Not sure what other implications running in that mode would have. Also, PostgreSQL doesn't support that statement as far as I know.

File: interface/src/main/scala/quasar/main/Quasar.scala (hunk 5 above)
Comment (drostron) [reply interface/src/main/scala/quasar/main/Quasar.scala]: Ideas?

File: interface/src/main/scala/quasar/main/Caching.scala (hunk 2 above)
Comment (drostron) [reply interface/src/main/scala/quasar/main/Caching.scala]: Just a starting point

File: core/src/main/scala/quasar/metastore/package.scala [hunk 16]
```diff
@@ -92,19 +96,68 @@ package object metastore {
...
+    })
+
   implicit val mountTypeMeta: Meta[MountType] = {
     import MountType._
     Meta[String].xmap[MountType](
       { case "view" => viewMount(); case "module" => moduleMount(); case fsType => fileSystemMount(FileSystemType(fsType)) },
       _.fold(_.value, "view", "module"))
   }
 
+  implicit val mountConfigComposite: Composite[MountConfig] =
```
[the remark this answers is not in the document]
Comment (drostron) [reply core/src/main/scala/quasar/metastore/package.scala]: A leftover addressing an undiagnosed NPE. Looks like it isn't necessary anymore.

File: web/src/main/scala/quasar/api/package.scala (hunk 13 above)
[the remark this answers is not in the document]
Comment (jedesah) [reply web/src/main/scala/quasar/api/package.scala]: Oh, ya, haha! That's just a comment that was left lying around in my pending comments that I totally forgot to delete since you had already fixed it! 😄

File: core/src/main/scala/quasar/fs/cache/ViewCache.scala (hunk 14 above)
[the remark this answers is not in the document]
Comment (jedesah) [reply core/src/main/scala/quasar/fs/cache/ViewCache.scala]: I think then that an `Option` or `EitherT` would be more appropriate. `Task` conveys that there is some kind of effect happening which is not the case

File: core/src/main/scala/quasar/metastore/MetaStoreAccess.scala (hunk 15 above)
[the remark this answers is not in the document]
Comment (jedesah) [reply core/src/main/scala/quasar/metastore/MetaStoreAccess.scala]: Ugh... so much for standards! 😞 I guess this is the best we can do then

File: interface/src/main/scala/quasar/main/Quasar.scala (hunk 4 above)
Comment (jedesah) [reply interface/src/main/scala/quasar/main/Quasar.scala]: No but seriously. I know this method is not used elsewhere in the codebase but it's tremendously useful for ad hoc testing in the `console`. Any reason why it should be removed?

File: interface/src/main/scala/quasar/main/Quasar.scala (hunk 5 above)
Comment (jedesah) [reply interface/src/main/scala/quasar/main/Quasar.scala]: I think `interp` or `coreInterp` would be fine. Not in love with them, but I think a slight improvement over `h`

File: interface/src/main/scala/quasar/main/Caching.scala (hunk 2 above)
Comment (jedesah) [reply interface/src/main/scala/quasar/main/Caching.scala]: The idea being that even if the machine has more than 4 cores it's unlikely to be faster to allocate more than 8 threads? I can believe it, but still curious what makes you think so. Perhaps a quick comment would be in order. If only so that future maintainers don't have to wonder if there is a really good reason why this was done this way if they ever feel the desire to change it

File: interface/src/main/scala/quasar/main/ViewCacheRefresh.scala [hunk 17]
```diff
@@ -70,16 +77,23 @@ object ViewCacheRefresh {
...
 
   // Naïve stale cache selection for the moment
   def selectCacheForRefresh[S[_]](implicit
     Q: QueryFile.Ops[S],
+    T: Timing.Ops[S],
     S0: ConnectionIO :<: S
   ): Q.transforms.CompExecM[Option[PathedViewCache]] =
-    lift(MetaStoreAccess.staleCachedViews ∘ (_.right[FileSystemError])) ∘ (_.headOption)
+    lift(T.timestamp ∘ (_.right[FileSystemError])) >>= {(ts: Instant) =>
+      println(s"ts: $ts")
```
Comment (jedesah) [inline interface/src/main/scala/quasar/main/ViewCacheRefresh.scala]: Looks like a leftover debug statement

File: web/src/main/scala/quasar/api/services/data.scala (hunk 10 above)
Comment (drostron) [reply web/src/main/scala/quasar/api/services/data.scala]: Agreed and meant to get back to this. Thanks for the reminder.

File: interface/src/main/scala/quasar/main/Quasar.scala (hunk 5 above)
Comment (jedesah) [reply interface/src/main/scala/quasar/main/Quasar.scala]: Fair enough. I didn't notice that actually. I have now become ambivalent

File: interface/src/main/scala/quasar/main/Caching.scala (hunk 2 above)
Comment (drostron) [reply interface/src/main/scala/quasar/main/Caching.scala]: This scales with `availableProcessors * 2` and if that happens to be below 8, rolls with 8 instead.
